packages/protocol/contracts/governance/Proposals.sol

```diff
@@ -0,0 +1,330 @@
+    } else if (previousVote == VoteValue.Yes) {
+      proposal.votes.yes = proposal.votes.yes.sub(weight);
+    } else if (previousVote == VoteValue.No) {
+      proposal.votes.no = proposal.votes.no.sub(weight);
+    }
+
+    // Add new vote.
+    if (currentVote == VoteValue.Abstain) {
+      proposal.votes.abstain = proposal.votes.abstain.add(weight);
+    } else if (currentVote == VoteValue.Yes) {
+      proposal.votes.yes = proposal.votes.yes.add(weight);
+    } else if (currentVote == VoteValue.No) {
+      proposal.votes.no = proposal.votes.no.add(weight);
+    }
+  }
+
+  /**
+   * @notice Executes the proposal, reverting if any transaction fails.
+   * @param proposal The proposal struct.
+   */
+  function execute(Proposal storage proposal) public {
+    for (uint256 i = 0; i < proposal.transactions.length; i = i.add(1)) {
+      // reverts proposal if any transaction fails
+      require(
+        externalCall(
+          proposal.transactions[i].destination,
+          proposal.transactions[i].value,
+          proposal.transactions[i].data.length,
+          proposal.transactions[i].data
+        )
+      );
+    }
+  }
+
+  /**
+   * @notice Computes the support ratio for a proposal with the quorum condition:
+   *   If the total participation (yes + no + abstain) is less than the critical
+   *   baseline, "no" votes are added to increase particiption to this level. The abstaining votes
+   *   are removed, leaving the "yes" and (potentially increased) "no" votes. The ratio of "yes"
+   *   votes to this value is returned.
+   * @param proposal The proposal struct.
+   * @param criticalBaseline The minimum participation at which "no" votes are not added.
+   * @return The support ratio with the quorum condition.
+   */
+  function adjustedSupport(
+    Proposal storage proposal,
+    int256 criticalBaseline
+  )
+    public
+    view
+    returns (int256)
+  {
+    if (proposal.votes.yes == 0) {
+      return 0;
+    }
+    int256 totalWeightFixed = toFixed(proposal.totalWeight);
+    int256 yesRatio = toFixed(proposal.votes.yes).divide(totalWeightFixed);
+    int256 abstainRatio = toFixed(proposal.votes.abstain).divide(totalWeightFixed);
+    int256 participation = toFixed(
+      proposal.votes.yes.add(proposal.votes.no).add(proposal.votes.abstain)
+    ).divide(totalWeightFixed);
+    int256 adjustedYesNoRatio = (
+      participation > criticalBaseline ? participation : criticalBaseline
+    ).subtract(abstainRatio);
+    int256 support = yesRatio.divide(adjustedYesNoRatio);
+    return support;
+  }
+
+  /**
+   * @notice Returns the stage of a dequeued proposal.
+   * @param proposal The proposal struct.
+   * @param stageDurations The durations of the dequeued proposal stages.
+   * @return The stage of the dequeued proposal.
+   */
+  function getDequeuedStage(
+    Proposal storage proposal,
+    StageDurations storage stageDurations
+  )
+    internal
+    view
+    returns (Stage)
+  {
```

Comment:
Should we require or assert that now >= proposal.timestamp? Or at least document that assumption?

Reply:
+1 on documenting, because that condition on its own isn't enough to guarantee that the proposal is dequeued. Will add a @dev comment noting the assumption